src/repr/type_binary.ml

```diff
@@ -19,112 +19,110 @@ open Staging
 open Utils
 
 module Encode = struct
-  let chars =
-    Array.init 256 (fun i -> Bytes.unsafe_to_string (Bytes.make 1 (Char.chr i)))
+  let unit () _byt off = off
 
-  let unit () _k = ()
-  let unsafe_add_bytes b k = k (Bytes.unsafe_to_string b)
-  let add_string s k = k s
-  let char c k = k chars.(Char.code c)
+  let add_string s byt off =
+    let ls = String.length s in
+    Bytes.blit_string s 0 byt off ls;
+    off + ls
 
-  let int8 i k =
-    assert (i < 256);
-    k chars.(i)
+  let add_bytes b byt off =
+    let lb = Bytes.length b in
+    Bytes.blit b 0 byt off lb;
+    off + lb
 
-  let int16 i =
-    let b = Bytes.create 2 in
-    Bytes.set_uint16_be b 0 i;
-    unsafe_add_bytes b
+  let char c byt off =
+    Bytes.set byt off c;
+    off + 1
 
-  let int32 i =
-    let b = Bytes.create 4 in
-    Bytes.set_int32_be b 0 i;
-    unsafe_add_bytes b
+  let byte n byt off = char (Char.chr (0xFF land n)) byt off
+  (* let unsafe_add_bytes b oc = IO.append_bytes oc b *)
 
-  let int64 i =
-    let b = Bytes.create 8 in
-    Bytes.set_int64_be b 0 i;
-    unsafe_add_bytes b
+  let int8 i byt off =
+    Bytes.set_uint8 byt off i;
+    off + 1
+
+  let int16 i byt off =
+    Bytes.set_uint16_be byt off i;
+    off + 2
+
+  let int32 (i : int32) byt off =
+    Bytes.set_int32_be byt off i;
+    off + 4
+
+  let int64 (i : int64) byt off =
+    Bytes.set_int64_be byt off i;
+    off + 8
 
   let float f = int64 (Int64.bits_of_float f)
   let bool b = char (if b then '\255' else '\000')
 
-  let int i k =
-    let rec aux n k =
-      if n >= 0 && n < 128 then k chars.(n)
-      else
-        let out = 128 lor (n land 127) in
-        k chars.(out);
-        aux (n lsr 7) k
-    in
-    aux i k
+  let rec aux n byt off =
```

Comment:
maybe rename `aux` into something more meaningful now that it is outside of `int`?